src/crimson/osd/pg.cc

```diff
@@ -970,50 +975,72 @@ seastar::future<> PG::wait_for_active()
     [[fallthrough]];
   case CEPH_OSD_OP_READ:
-    return backend->read(oi,
+    return backend->read(os.oi,
                          op.extent.offset,
                          op.extent.length,
                          op.extent.truncate_size,
                          op.extent.truncate_seq,
-                         op.flags).then([osd_op](bufferlist bl) {
-      osd_op->rval = bl.length();
-      osd_op->outdata = std::move(bl);
+                         op.flags).then([&osd_op](bufferlist bl) {
+      osd_op.rval = bl.length();
+      osd_op.outdata = std::move(bl);
       return seastar::now();
     });
-  default:
+  case CEPH_OSD_OP_WRITE:
+    // TODO: handle write separately. For `rados bench write` the fall-
+    // through path somehow works but this is really nasty.
+    [[fallthrough]];
+  case CEPH_OSD_OP_WRITEFULL:
+    // XXX: os = backend->write(std::move(os), ...) instead?
+    return backend->writefull(os, osd_op, txn);
+  case CEPH_OSD_OP_SETALLOCHINT:
     return seastar::now();
+  default:
+    throw std::runtime_error(
+      fmt::format("op '{}' not supported", ceph_osd_op_name(op.op)));
   }
 }
 
 seastar::future<Ref<MOSDOpReply>> PG::do_osd_ops(Ref<MOSDOp> m)
 {
-  // todo: issue requests in parallel if they don't write,
-  // with writes being basically a synchronization barrier
-  return seastar::do_with(std::move(m), [this](auto& m) {
-    return seastar::do_for_each(begin(m->ops), end(m->ops),
-                                [m,this](OSDOp& osd_op) {
-      const auto oid = (m->get_snapid() == CEPH_SNAPDIR ?
-                        m->get_hobj().get_head() :
-                        m->get_hobj());
-      return backend->get_object(oid).then([&osd_op,this](auto oi) {
-        return do_osd_op(*oi, &osd_op);
-      }).handle_exception_type([&osd_op](const object_not_found&) {
-        osd_op.rval = -ENOENT;
-        throw;
+  return seastar::do_with(std::move(m), ceph::os::Transaction{},
+                          [this](auto& m, auto& txn) {
+    const auto oid = m->get_snapid() == CEPH_SNAPDIR ? m->get_hobj().get_head()
+                                                     : m->get_hobj();
+    return backend->get_object_state(oid).then([m,&txn,this](auto os) {
+      // TODO: issue requests in parallel if they don't write,
+      // with writes being basically a synchronization barrier
+      return seastar::do_for_each(std::begin(m->ops), std::end(m->ops),
+                                  [m,&txn,this,os](OSDOp& osd_op) {
+        return do_osd_op(*os, osd_op, txn);
+      }).then([m,&txn,this,os=std::move(os)] {
+        // XXX: the entire lambda can be scheduled conditionally
+        // XXX: I'm not txn.empty() is what we want here
+        return !txn.empty() ? backend->store_object_state(os, *m, txn)
+                            : seastar::now();
       });
+    }).then([&] {
+      return txn.empty() ? seastar::now()
+                         : backend->submit_transaction(std::move(txn));
```

Comment:
Might be reasonable to unify `submit_transaction` and `store_object_state`. Alternatively we could just make the former one returning `void` instead of `seastar::future<>` as it seems to be overkill.